File: q70-q79/q79/q79.py
class Solution:
    def exist(self,board, word):
        row = len(board)
        col = len(board[0])
        wlen = len(word)
        i, j = 0, 0
        def find(i, j,index):
            if index >= wlen:
                return True
            temp=board[i][j]
            board[i][j]=None
            if i + 1 < row and j < col:
                if board[i + 1][j] == word[index]:
                    if find(i + 1, j,index + 1):
                        return True
            if i - 1 >= 0 and j < col:
                if board[i - 1][j] == word[index]:
                    if find(i - 1, j,  index + 1):
                        return True
            if i < row and j + 1 < col:
                if board[i][j + 1] == word[index]:
                    if find(i, j + 1,index + 1):
                        return True
            if i < row and j - 1 >= 0:
                if board[i][j - 1] == word[index]:
                    if find(i, j - 1,index + 1):
                        return True
            board[i][j]=temp
            return False


        for i in range(row):
            for j in range(col):
                if str(board[i][j]) == word[0]:
                    if find(i,j,1):
                        return True

        return False

File: q70-q79/q79/q79.cpp
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        if (board.empty() || board[0].empty() || word.empty()) {
            return false;
        }
        
        int row = board.size(), col = board[0].size();
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (word[0] == board[i][j] && dfs(board, word, i, j, 0)) {
                    return true;
                }
            }
        }
        return false;
    }
    
private:
    const vector<vector<int>> dirs {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
    
    bool dfs(vector<vector<char>>& board, string word, int row, int col, int index) {
        if (index == word.size()) {
            return true;
        }
        
        if (row < 0 || col < 0 || row >= board.size() || col >= board[0].size() || board[row][col] != word[index]) {
            return false;
        }
        
        char old = board[row][col];
        board[row][col] = '$';
        
        // all branches
        for (const auto& dir : dirs) {
            if (dfs(board, word, row + dir[0], col + dir[1], index + 1)) {
                board[row][col] = old;
                return true;
            }
        }
        
        board[row][col] = old;
        return false;
    }
};
